File: packages/python/prompt_sanitizer/synthetic.py
"""
Synthetic Replacement Engine.

Generates realistic fake values per EntityType using the Faker library.
If Faker is not installed, falls back to sequential placeholder tokens
like ``[EMAIL_1]``, ``[PERSON_2]``, etc.

The vault ensures that within a session the same original always maps
to the same fake value (determinism).
"""
from __future__ import annotations

from collections import defaultdict

from .entities import EntityType

try:
    from faker import Faker as _Faker  # type: ignore[import]
    _HAS_FAKER = True
except ImportError:
    _HAS_FAKER = False


class SyntheticEngine:
    """
    Generates context-appropriate fake replacement values.

    Usage::

        engine = SyntheticEngine(locale="en_US")
        fake_email = engine.generate(EntityType.EMAIL, "john.doe@acme.com")
        # → "xavier.chen@mailnull.net"
    """

    def __init__(self, locale: str = "en_US") -> None:
        self._locale = locale
        self._fake = _Faker(locale) if _HAS_FAKER else None
        # Per-type counters for placeholder fallback
        self._counters: dict[EntityType, int] = defaultdict(int)

    # ── Public ───────────────────────────────────────────────────────────────

    def generate(self, entity_type: EntityType, original: str = "") -> str:
        """Return a fake value for *entity_type*."""
        if self._fake is not None:
            return self._faker_value(entity_type, original)
        return self._placeholder(entity_type)

    # ── Faker-backed generation ───────────────────────────────────────────────

    def _faker_value(self, entity_type: EntityType, original: str) -> str:
        f = self._fake
        assert f is not None

        match entity_type:
            case EntityType.PERSON:
                return f.name()
            case EntityType.EMAIL:
                return f.ascii_safe_email()
            case EntityType.PHONE:
                return f.phone_number()
            case EntityType.SSN:
                # Generate a structurally valid but fake SSN
                return f"{f.random_int(100, 899):03d}-{f.random_int(10, 99):02d}-{f.random_int(1000, 9999):04d}"
            case EntityType.CREDIT_CARD:
                return _fake_luhn_card(f)
            case EntityType.IBAN:
                return f"GB{f.random_int(10,99):02d}MOCK{f.random_int(10000000, 99999999):08d}{f.random_int(100000000000, 999999999999):012d}"
            case EntityType.IP_ADDRESS:
                return f"{f.random_int(1,254)}.{f.random_int(0,255)}.{f.random_int(0,255)}.{f.random_int(1,254)}"
            case EntityType.MAC_ADDRESS:
                return ":".join(f"{f.random_int(0,255):02x}" for _ in range(6))
            case EntityType.URL:
                return f"https://{f.domain_name()}/{f.uri_path()}"
            case EntityType.ADDRESS:
                return f.address().replace("\n", ", ")
            case EntityType.ZIP_CODE:
                return f.postcode()
            case EntityType.DATE:
                return f.date(pattern="%m/%d/%Y")
            case EntityType.DATE_OF_BIRTH:
                return f.date_of_birth(minimum_age=18, maximum_age=80).strftime("%m/%d/%Y")
            case EntityType.CRYPTO_ADDRESS:
                # Fake-looking ETH address (not valid on chain)
                return "0x" + "".join(f"{f.random_int(0,15):x}" for _ in range(40))
            case EntityType.PASSPORT:
                return f"{chr(f.random_int(65,90))}{f.random_int(10000000, 99999999)}"
            case EntityType.DRIVING_LICENSE:
                return f"{chr(f.random_int(65,90))}{f.random_int(100000, 999999)}"
            case EntityType.API_KEY:
                return "sk-" + "".join(
                    f.random_element("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789")
                    for _ in range(48)
                )
            case EntityType.JWT:
                return "eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJSRURBQ1RFRCJ9.REDACTED_SIGNATURE"
            case EntityType.BEARER_TOKEN:
                return "REDACTED_" + "".join(
                    f.random_element("ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789") for _ in range(16)
                )
            case EntityType.AWS_ACCESS_KEY:
                return "AKIAIOSFODNN7EXAMPLE"
            case EntityType.AWS_SECRET_KEY:
                return "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
            case EntityType.PRIVATE_KEY:
                return "-----BEGIN PRIVATE KEY-----\nREDACTED\n-----END PRIVATE KEY-----"
            case EntityType.DB_CONNECTION:
                return f"postgresql://user:password@localhost:5432/{f.word()}"
            case _:
                return self._placeholder(entity_type)

    # ── Placeholder fallback (no Faker) ──────────────────────────────────────

    def _placeholder(self, entity_type: EntityType) -> str:
        self._counters[entity_type] += 1
        return f"[{entity_type.value}_{self._counters[entity_type]}]"


# ---------------------------------------------------------------------------
# Luhn-valid fake credit card generator
# ---------------------------------------------------------------------------

def _fake_luhn_card(f: object) -> str:  # type: ignore[type-arg]
    """Generate a 16-digit Luhn-valid fake card number (Visa prefix)."""
    import random
    prefix = [4]  # Visa
    digits = prefix + [random.randint(0, 9) for _ in range(14)]
    # Calculate check digit
    odd = digits[-1::-2]
    even = digits[-2::-2]
    total = sum(odd) + sum(sum(divmod(d * 2, 10)) for d in even)
    check = (10 - (total % 10)) % 10
    digits.append(check)
    raw = "".join(str(d) for d in digits)
    return f"{raw[:4]} {raw[4:8]} {raw[8:12]} {raw[12:]}"

File: packages/ruby/lib/prompt_sanitizer/synthetic.rb
# frozen_string_literal: true

module PromptSanitizer
  # Generates realistic fake replacement values per EntityType.
  #
  # When the +faker+ gem is installed, each call produces a
  # contextually appropriate fake (names, emails, IPs, …).
  # Without faker the engine falls back to sequential placeholder
  # tokens: +[EMAIL_1]+, +[PERSON_2]+, etc.
  #
  # Determinism within a session is guaranteed by the Vault: the
  # same original value always receives the same token/fake because
  # Session#anonymize only calls +generate+ once per unique original.
  #
  # Usage:
  #
  #   engine = SyntheticEngine.new(locale: "en_US")
  #   engine.generate(:email, "john@acme.com")   # => "xavier@mailnull.net"
  #   engine.generate(:person, "Alice Smith")    # => "Carlos Rivera"
  class SyntheticEngine
    CHARS_ALPHA_NUM = ("a".."z").to_a + ("A".."Z").to_a + ("0".."9").to_a

    begin
      require "faker"
      HAS_FAKER = true
    rescue LoadError
      HAS_FAKER = false
    end

    # @param locale [String] BCP-47 locale tag forwarded to Faker (e.g. "en", "fr", "de")
    def initialize(locale: "en")
      @locale   = locale
      @counters = Hash.new(0) # entity_type Symbol → Integer
      if HAS_FAKER
        Faker::Config.locale = locale
      end
    end

    # Returns a fake replacement string for +entity_type+.
    #
    # @param entity_type [Symbol]  one of the EntityType constants
    # @param _original   [String]  original text (unused here; determinism via Vault)
    # @return [String]
    def generate(entity_type, _original = "")
      if HAS_FAKER
        faker_value(entity_type)
      else
        placeholder(entity_type)
      end
    end

    # Force a placeholder token regardless of faker availability.
    # Used by Session when the replacement must survive round-trips.
    def placeholder(entity_type)
      @counters[entity_type] += 1
      "[#{entity_type.to_s.upcase}_#{@counters[entity_type]}]"
    end

    def reset!
      @counters.clear
    end

    private

    # rubocop:disable Metrics/MethodLength, Metrics/CyclomaticComplexity
    def faker_value(entity_type) # rubocop:disable Metrics/AbcSize
      case entity_type
      when :person
        Faker::Name.name
      when :email
        Faker::Internet.email
      when :phone
        Faker::PhoneNumber.phone_number
      when :ssn
        "#{rand(100..899).to_s.rjust(3, "0")}-" \
          "#{rand(10..99).to_s.rjust(2, "0")}-" \
          "#{rand(1000..9999)}"
      when :credit_card
        fake_luhn_card
      when :iban
        "GB#{rand(10..99)}MOCK" \
          "#{rand(10_000_000..99_999_999)}" \
          "#{rand(100_000_000_000..999_999_999_999)}"
      when :ip_address
        "#{rand(1..254)}.#{rand(0..255)}.#{rand(0..255)}.#{rand(1..254)}"
      when :mac_address
        Array.new(6) { rand(0..255).to_s(16).rjust(2, "0") }.join(":")
      when :url
        "https://#{Faker::Internet.domain_name}/#{Faker::Internet.slug}"
      when :address
        Faker::Address.full_address.tr("\n", ", ")
      when :zip_code
        Faker::Address.postcode
      when :date
        Faker::Date.between(from: "1990-01-01", to: "2020-12-31").strftime("%m/%d/%Y")
      when :date_of_birth
        Faker::Date.birthday(min_age: 18, max_age: 80).strftime("%m/%d/%Y")
      when :crypto_address
        "0x" + Array.new(40) { rand(0..15).to_s(16) }.join
      when :passport
        "#{("A".."Z").to_a.sample}#{rand(10_000_000..99_999_999)}"
      when :driving_license
        "#{("A".."Z").to_a.sample}#{rand(100_000..999_999)}"
      when :organization
        Faker::Company.name
      when :location
        Faker::Address.city
      when :api_key
        "sk-" + Array.new(48) { CHARS_ALPHA_NUM.sample }.join
      when :jwt, :jwt_token
        "eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJSRURBQ1RFRCJ9.REDACTED_SIGNATURE"
      when :bearer_token, :oauth_token
        "REDACTED_" + Array.new(16) { (("A".."Z").to_a + ("0".."9").to_a).sample }.join
      when :aws_access_key
        "AKIA" + Array.new(16) { (("A".."Z").to_a + ("0".."9").to_a).sample }.join
      when :aws_secret_key
        Array.new(40) { (("a".."z").to_a + ("A".."Z").to_a + ("0".."9").to_a + ["+", "/"]).sample }.join
      when :private_key
        "-----BEGIN PRIVATE KEY-----\nREDACTED\n-----END PRIVATE KEY-----"
      when :db_connection, :database_url
        "postgresql://user:password@localhost:5432/#{Faker::Lorem.word}"
      when :secret_key, :password
        placeholder(entity_type)
      else
        placeholder(entity_type)
      end
    end
    # rubocop:enable Metrics/MethodLength, Metrics/CyclomaticComplexity

    # Generates a Luhn-valid 16-digit fake Visa card number.
    def fake_luhn_card
      digits = [4] + Array.new(14) { rand(0..9) }

      # Calculate check digit — double digits at even indices (0,2,4,...,14)
      # so that in the final 16-digit number they sit at even positions from
      # the right (2,4,...,16) as required by the Luhn algorithm.
      sum = digits.each_with_index.sum do |d, i|
        if i.even?
          doubled = d * 2
          doubled > 9 ? doubled - 9 : doubled
        else
          d
        end
      end
      check = (10 - (sum % 10)) % 10
      digits << check

      raw = digits.join
      "#{raw[0, 4]} #{raw[4, 4]} #{raw[8, 4]} #{raw[12, 4]}"
    end
  end
end
